common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,408 @@
+package txmgr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"gopkg.in/guregu/null.v4"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+)
+
+var (
+	// ErrInvalidChainID is returned when the chain ID is invalid
+	ErrInvalidChainID = errors.New("invalid chain ID")
+	// ErrTxnNotFound is returned when a transaction is not found
+	ErrTxnNotFound = errors.New("transaction not found")
+	// ErrExistingIdempotencyKey is returned when a transaction with the same idempotency key already exists
+	ErrExistingIdempotencyKey = errors.New("transaction with idempotency key already exists")
+	// ErrAddressNotFound is returned when an address is not found
+	ErrAddressNotFound = errors.New("address not found")
+	// ErrSequenceNotFound is returned when a sequence is not found
+	ErrSequenceNotFound = errors.New("sequence not found")
+	// ErrCouldNotGetReceipt is the error string we save if we reach our finality depth for a confirmed transaction without ever getting a receipt
+	// This most likely happened because an external wallet used the account for this nonce
+	ErrCouldNotGetReceipt = errors.New("could not get receipt")
+)
+
+type InMemoryStore[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+] struct {
+	lggr    logger.SugaredLogger
+	chainID CHAIN_ID
+
+	keyStore txmgrtypes.KeyStore[ADDR, CHAIN_ID, SEQ]
+	txStore  txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+
+	addressStatesLock sync.RWMutex
+	addressStates     map[ADDR]*AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+}
+
+// NewInMemoryStore returns a new InMemoryStore
+func NewInMemoryStore[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](
+	ctx context.Context,
+	lggr logger.SugaredLogger,
+	chainID CHAIN_ID,
+	keyStore txmgrtypes.KeyStore[ADDR, CHAIN_ID, SEQ],
+	txStore txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE],
+	config txmgrtypes.InMemoryStoreConfig,
+) (*InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE], error) {
+	ms := InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		lggr:     lggr,
+		chainID:  chainID,
+		keyStore: keyStore,
+		txStore:  txStore,
+
+		addressStates: map[ADDR]*AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{},
+	}
+
+	maxUnstarted := int(config.MaxQueued())
```

Comment:
Left a related comment in the Step 2 PR but if we update `NewAddressState` to take `uint64` for `maxUnstarted`, we could skip this cast.